### PR TITLE
Added Deep Contextual Bandits

### PR DESCRIPTION
**Pull request by threewisemonkeys-as**
Addresses  #173 

Added the following DCBs till now -
1. Linear Posterior Sampling (Bayesian Linear Regression on labels)
2. Neural Linear Posterior Sampling (Bayesian Linear Regression on labels using latent representation by NN)
Both are taken from the [paper](https://arxiv.org/abs/1802.09127) from Google Brain

I plan to add the other methods discussed. Just drafted this to discuss organisation of code. I have currently added all code to `classical/bandit` with the data based bandit added to `bandit/data_bandits` and the agents to `bandit/dcb_agents`  Issues with this are -
1. Not all of the methods are necessarily deep. We can therefore just have a `bandits/agents` dir for all the policies / agents and a `bandit/bandits` for all the bandit implementations
2. All of bandits are based on data which is usually a large text file. Without this, the bandits are non - functional. Should we include the data files in the repo (I have done this temporarily) or is there some other solution?
3. While the implementations of the agents are functional, they are not converging. I am currently working on fixing this.

### CODE REVIEW

**Comment by threewisemonkeys-as** [comment]
For the data, I was thinking of adding a script to download it, like in pytorch's builtin datasets

**Comment by Sharad24** [comment]
Can you upload the plots for the testbed?  I'll try to take a look at the other dataset today

**Comment by Sharad24** [comment]
Let me know if things don't work, lets try and get some working by Wednesday

**Comment by threewisemonkeys-as** [comment]
## BootstrapNeuralAgent-on-BernoulliCB
BootstrapNeuralAgent-on-BernoulliCB

## BootstrapNeuralAgent-on-CovertypeDataBandit
BootstrapNeuralAgent-on-CovertypeDataBandit

## FixedAgent-on-BernoulliCB.png
FixedAgent-on-BernoulliCB

## FixedAgent-on-CovertypeDataBandit
FixedAgent-on-CovertypeDataBandit

## LinearPosteriorAgent-on-BernoulliCB
LinearPosteriorAgent-on-BernoulliCB

## LinearPosteriorAgent-on-CovertypeDataBandit
LinearPosteriorAgent-on-CovertypeDataBandit

## NeuralNoiseSamplingAgent-on-BernoulliCB
NeuralNoiseSamplingAgent-on-BernoulliCB

## NeuralNoiseSamplingAgent-on-CovertypeDataBandit 
NeuralNoiseSamplingAgent-on-CovertypeDataBandit

## NeuralGreedyAgent-on-BernoulliCB 
NeuralGreedyAgent-on-BernoulliCB

## NeuralGreedyAgent-on-CovertypeDataBandit 
NeuralGreedyAgent-on-CovertypeDataBandit

## NeuralLinearPosteriorAgent-on-BernoulliCB 
NeuralLinearPosteriorAgent-on-BernoulliCB

## NeuralLinearPosteriorAgent-on-CovertypeDataBandit 
NeuralLinearPosteriorAgent-on-CovertypeDataBandit

## VariationalAgent-on-BernoulliCB 
VariationalAgent-on-BernoulliCB

## VariationalAgent-on-CovertypeDataBandit 
VariationalAgent-on-CovertypeDataBandit

**Comment by threewisemonkeys-as** [comment]
@Sharad24  the ones that seem to be working right now (at least somewhat) are -

1. NeuralGreedy
2. NeuralBootstrap
3. LinearPosterior
4. Variational (but only on covertype)

**Comment by Sharad24** [comment]
Could you put these in one plot with a higher DPI? Seems very hard to compare separately

**Comment by threewisemonkeys-as** [comment]
DCBs-on-CovertypeDataBandit

**Comment by threewisemonkeys-as** [comment]
Neural LinPos and Neural Noise arent learning at all while Variational is a bit subpar. The others seems to be working fine. The issue is most like with the bayesian neural net I created.

**Comment by threewisemonkeys-as** [comment]
DCBs-on-BernoulliCB

**Comment by Sharad24** [comment]
@threewisemonkeys-as Whats the status here? Lets take stock and merge at least what we have working since this is a very big PR

**Comment by threewisemonkeys-as** [comment]
I've gotten Noise Sampling to also work by adding LR annealing and gradient clipping, only issue is Neural Linear which is not working. 

I'll add docstrings to the current code and open for review.

**Comment by threewisemonkeys-as** [comment]
@Sharad24 have a look.  I think we should merge this. Further changes can be made in separate PRs then.

File: genrl/deep/bandit/data_bandits/mushroom_bandit.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+from pathlib import Path
+from typing import Tuple, Union
+
+import numpy as np
+import pandas as pd
+import torch
+
+from .data_bandit import DataBasedBandit, download_data
```
Comment (Sharad24) [inline genrl/deep/bandit/data_bandits/mushroom_bandit.py:8]: We're moving away from relative paths. So you'd have to change this at a lot of places now.

File: genrl/deep/bandit/dcb_agents/bootstrap_neural.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+import numpy as np
+import torch
+
+from ..data_bandits import DataBasedBandit
+from .common import NeuralBanditModel, TransitionDB
+from .dcb_agent import DCBAgent
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+dtype = torch.float
```
Comment (Sharad24) [inline genrl/deep/bandit/dcb_agents/bootstrap_neural.py:11]: Are these variables necessary?

File: genrl/deep/bandit/dcb_agents/bootstrap_neural.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
+from typing import List, Optional
+
+import numpy as np
+import torch
+
+from ..data_bandits import DataBasedBandit
+from .common import NeuralBanditModel, TransitionDB
+from .dcb_agent import DCBAgent
```
Comment (Sharad24) [inline genrl/deep/bandit/dcb_agents/bootstrap_neural.py:8]: Relative paths

File: genrl/deep/bandit/dcb_agents/common.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+from typing import List, Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import dropout, dropout_
+from torch.cuda import init
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+dtype = torch.float
```
Comment (Sharad24) [inline genrl/deep/bandit/dcb_agents/common.py:11]: Same comment about the necessity of these variables? My issue really is that, if even one of these files is imported somewhere, these variables are implicitly imported as well

File: genrl/deep/bandit/dcb_agents/dcb_agent.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+from typing import Optional
+
+import torch
+
+from ..data_bandits import DataBasedBandit
```
Comment (Sharad24) [inline genrl/deep/bandit/dcb_agents/dcb_agent.py:5]: Relative paths

**Comment by Sharad24** [comment]
Also, can you take a look at the LGTM alerts and Code Climate Issues(28!?) ?

File: genrl/deep/bandit/data_bandits/mushroom_bandit.py (hunk 1 above)
Comment (threewisemonkeys-as) [reply genrl/deep/bandit/data_bandits/mushroom_bandit.py:8]: Yeah ok, will change those

File: genrl/deep/bandit/dcb_agents/bootstrap_neural.py (hunk 2 above)
Comment (threewisemonkeys-as) [reply genrl/deep/bandit/dcb_agents/bootstrap_neural.py:11]: I'll add device as an argument and get rid of dtype

**Comment by threewisemonkeys-as** [comment]
Most of the code climate issues are that `__init__` functions have too many args and some cognitive complexity issues. Ill try to simplify some functions but dont know what to do about the too many arguments issue

**Comment by sampreet-arthi** [comment]
You can try using *args, **kwargs to reduce no of variables being used.

**Comment by threewisemonkeys-as** [comment]
Wouldn't that be even more confusing to the user as opposed to many arguments? Could we use something like a hyperparam object like they do in tf ?

**Comment by sampreet-arthi** [comment]
I personally find the hyperparam object very unpythonic, it's even more confusing because all the arguments are specified elsewhere if I'm not wrong. It would be simpler if we had args and kwargs and wrote some neat docs instead.

**Comment by threewisemonkeys-as** [comment]
Hmm ok, I'll update this also. Btw are we planning on doing the same for the `deep/agents` which have similarly large list of args.

**Comment by sampreet-arthi** [comment]
> Hmm ok, I'll update this also. Btw are we planning on doing the same for the `deep/agents` which have similarly large list of args.

@Sharad24 @ajaysub110 @Het-Shah Can we do this? Check out #198 for an example of how it can be done. It really makes the code a lot shorter and neater (at least 20 lines of code can be removed for most algos)

**Comment by Sharad24** [comment]
Yup. This was an issue planned long before #152

**Comment by threewisemonkeys-as** [comment]
Resolved most of the complexity issues, Most of the ones remaining are in `examples/deep_cb.py` . Not really sure how I can fix those in the `bandits/agents/cb_agents/common/` though.

File: genrl/bandit/agents/cb_agents/common/bayesian.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                if kl:
+                    kl_val += torch.sum(
+                        b_dist.log_prob(b)
+                        - torch.distributions.Normal(0, 0.1).log_prob(b)
+                    )
+            else:
+                b = 0.0
+                # b_logprob = None
+
+        return F.linear(x, w, b), kl_val
```
Comment (sampreet-arthi) [inline genrl/bandit/agents/cb_agents/common/bayesian.py:86]: The codeclimate issue can be resolved by putting in as few if else statements as possible. This is a pretty long block of code with lots of nested if statements. Would be better if you could cover this in fewer lines with lesser if statements. (atleast try to go without nested). Inline if else assignments may help too.

File: genrl/bandit/agents/cb_agents/common/neural.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    Attributes:
+        use_dropout (bool): Indicated whether or not dropout should be used in forward pass.
+    """
+
+    def __init__(self, **kwargs):
+        super(NeuralBanditModel, self).__init__()
+        self.context_dim = kwargs.get("context_dim")
+        self.hidden_dims = kwargs.get("hidden_dims")
+        self.n_actions = kwargs.get("n_actions")
```
Comment (sampreet-arthi) [inline genrl/bandit/agents/cb_agents/common/neural.py:34]: Don't use kwargs here, you can use the actual variable names and and use kwargs only if needed

File: genrl/bandit/agents/cb_agents/bootstrap_neural.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            NeuralBanditModel(
+                context_dim=self.context_dim,
+                hidden_dims=kwargs.get("hidden_dims", [50, 50]),
+                n_actions=self.n_actions,
+                init_lr=kwargs.get("init_lr", 0.1),
+                max_grad_norm=kwargs.get("max_grad_norm", 0.5),
+                lr_decay=kwargs.get("lr_decay", 0.5),
+                lr_reset=kwargs.get("lr_reset", True),
+                dropout_p=kwargs.get("dropout_p", None),
+            )
```
Comment (sampreet-arthi) [inline genrl/bandit/agents/cb_agents/bootstrap_neural.py:52]: Not necessary? You could just send **kwargs instead.

File: genrl/bandit/agents/cb_agents/linpos.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+                        torch.distributions.MultivariateNormal(
+                            self.mu[i], var[i] * self.cov[i]
+                        ).sample()
+                        for i in range(self.n_actions)
+                    ]
+                )
+                .to(self.device)
+                .to(torch.float)
+            )
+        except RuntimeError as e:  # noqa F841
```
Comment (sampreet-arthi) [inline genrl/bandit/agents/cb_agents/linpos.py:96]: Seems very fishy to me. Why would you except a generic RuntimeError? Maybe it would be better to initiate default variables to pass instead of `self.mu[i], var[i] * self[i]`?

File: genrl/bandit/agents/cb_agents/bootstrap_neural.py (hunk 8 above)
Comment (threewisemonkeys-as) [reply genrl/bandit/agents/cb_agents/bootstrap_neural.py:52]: THought it would be better to be more explicit here

File: genrl/bandit/agents/cb_agents/linpos.py (hunk 9 above)
Comment (threewisemonkeys-as) [reply genrl/bandit/agents/cb_agents/linpos.py:96]: Yeah, I'll change this to a more specific error

File: genrl/bandit/agents/cb_agents/bootstrap_neural.py (hunk 8 above)
Comment (sampreet-arthi) [reply genrl/bandit/agents/cb_agents/bootstrap_neural.py:52]: Then it would be better not to use kwargs.get() 

Pass the proper arguments instead

File: genrl/bandit/agents/cb_agents/common/neural.py (hunk 7 above)
Comment (threewisemonkeys-as) [reply genrl/bandit/agents/cb_agents/common/neural.py:34]: Then I get code climate issue for too many args

File: genrl/bandit/agents/cb_agents/bootstrap_neural.py (hunk 8 above)
Comment (threewisemonkeys-as) [reply genrl/bandit/agents/cb_agents/bootstrap_neural.py:52]: That would just unncessarily add more lines to store the vars seperately right?

**Comment by threewisemonkeys-as** [comment]
I have added support for training with cli. So something like this
`python -m genrl.bandit.main -a neural-linpos -b covertype -t 15000 --logdir ./logs`

**Comment by threewisemonkeys-as** [comment]
This is pretty much done from my end. The remaining code climate issues are pretty minor.

Anything else to do here @Sharad24 @sampreet-arthi  or can we merge?

**Comment by threewisemonkeys-as** [comment]
I have removed the bandits in the old location. (They are now shifted to `/bandit`)

**Comment by sampreet-arthi** [comment]
Maybe once the ci checks out after you resolve the merge conflicts?

**Comment by Sharad24** [comment]
Anything else to add? Merge at will
